passes/cmds/logger.cc

```diff
@@ -158,12 +159,40 @@ struct LoggerPass : public Pass {
 					log_cmd_error("Expected error message occurrences must be 1 !\n");
 				log("Added regex '%s' for warnings to expected %s list.\n", pattern.c_str(), type.c_str());
 				try {
-					if (type=="error")
-						log_expect_error.push_back(std::make_pair(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count)));
-					else if (type=="warning")
-						log_expect_warning.push_back(std::make_pair(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count)));
-					else
-						log_expect_log.push_back(std::make_pair(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count)));
+					if (type=="error") {
+						auto it = log_expect_error.begin();
+						auto ie = log_expect_error.end();
+						for (; it != ie; it++)
+							if (it->second.pattern == pattern) {
+								it->second.expected_count = count;
+								break;
+							}
+						if (it == ie)
+							log_expect_error.emplace_back(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count));
+					}
+					else if (type=="warning") {
+						auto it = log_expect_warning.begin();
+						auto ie = log_expect_warning.end();
+						for (; it != ie; it++)
+							if (it->second.pattern == pattern) {
+								it->second.expected_count = count;
+								break;
+							}
+						if (it == ie)
+							log_expect_warning.emplace_back(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count));
+					}
+					else if (type=="log") {
+						auto it = log_expect_log.begin();
+						auto ie = log_expect_log.end();
+						for (; it != ie; it++)
+							if (it->second.pattern == pattern) {
+								it->second.expected_count = count;
+								break;
+							}
+						if (it == ie)
+							log_expect_log.emplace_back(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count));
+					}
```

Comment:
Does something like this not work because of difficulties comparing the elements of the `std::pair`?

Make `log_expect_error`, etc. `std::set` instead of `std::vector` and:

```suggestion

					if (type == "error")
						log_expect_error.insert(std::make_pair(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count)));
					else if (type == "warning")
						log_expect_warning.insert(std::make_pair(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count)));
					else if (type == "log")
						log_expect_log.insert(std::make_pair(YS_REGEX_COMPILE(pattern), LogExpectedItem(pattern, count)));
```

Reply:
It's unclear whether all instances of `YS_REGEX_COMPILE` resolve to objects that are less-than-comparable. Note that I am walking over `log_expect_{error,warning,log}` and equality-checking `<pair>.second.pattern`.

Reply:
I see, because there is some uncertainty about what exactly `YS_REGEX_COMPILE` is (at least until #1772 is closed), it's better to take the safe route.  But it would be good to check, because this is significantly more verbose.  And even if there is a problem with comparison, the `log_expect_*` should still be `std::set`, because that more closely expresses the design intent: Each element should be unique.